packages/teams-js/src/internal/communication.ts

```diff
@@ -257,6 +264,90 @@ export function sendMessageToParentAsync<T>(
   });
 }
 
+type SimpleType = string | number | boolean | null | undefined | SimpleType[];
+
+function isSerializable(arg: unknown): arg is ISerializable {
+  return (
+    arg !== undefined &&
+    arg !== null &&
+    (arg as ISerializable).serialize !== undefined &&
+    typeof (arg as ISerializable).serialize === 'function'
+  );
+}
+
+function serializeItemArray(items: (SimpleType | ISerializable)[]): unknown[] {
+  return items.map((item) => {
+    if (isSerializable(item)) {
+      return item.serialize();
+    } else {
+      return item;
+    }
+  });
+}
+
+/**
+ * Call a function in the host and receive a response. If the host returns an {@link SdkError} instead of a normal response, this function will throw a new Error containing the SdkError's information
+ *
+ * @param apiVersionTag A unique tag used to identify the API version for telemetry purposes.
+ * @param functionName The function name to call in the host.
+ * @param responseHandler When the host responds, this handler will validate and deserialize the response.
+ * @param args A collection of data to pass to the host. This data must be an array of either simple types or objects that implement {@link ISerializable}.
+ * @param errorChecker This optional property can be used to override the default ErrorChecking this function uses to decide whether to throw the host response as a new Error. Specify this is your function needs to do any logic verifying that the object received is an error that goes beyond the logic found in {@link isSdkError}.
+ *
+ * @returns The response received from the host after deserialization.
+ *
+ * @throws An Error containing the SdkError information ({@link SdkError.errorCode} and {@link SdkError.message}) if the host returns an SdkError, or an Error if the response from the host is an unexpected format.
+ */
+export async function callFunctionInHostAndHandleResponse<ReceivedFromHost, DeserializedFromHost>(
+  apiVersionTag: string,
+  functionName: string,
+  responseHandler: ResponseHandler<ReceivedFromHost, DeserializedFromHost>,
+  args: (SimpleType | ISerializable)[],
+  errorChecker?: (response: unknown) => response is SdkError,
```

Comment:
>errorChecker [](http://example.com/codeflow?start=2&length=12)

What's an example use case for this? Something about this strikes me as a little funny. It seems like this code should always be able to determine itself if the response object is an `SdkError` (e.g., we already have the `isSdkError` function).

I'm looking at the example in the `authenticateWithOauth` where it provides a custom checker that checks if the object is an `InvokeError` -- and then it _essentially_ lies by saying "I just identified this as an `InvokeError` but yeah, it's totally an `SdkError`"

I get that those two types are strikingly similar but they aren't actually the same. 

Let me think about why this seems funky to me.

Reply:
It seems like what you actually want this function to be is something like "If you aren't using `SdkError`, then this function's job is to return an error object containing an error code and a message field that is a transformation of your custom error type" -- which isn't the same thing as saying that this custom object IS an `SdkError`, because they aren't. At least in the case of `InvokeError` the interpretation of the `errorCode` field should be different.

Maybe you should create a new interface like `ICommonError` and this parameter should be `transformErrorToCommonError?: (response: unknown) => ICommonError` and then your code below would look like this:

```
  if ((transformErrorToCommonError&& transformErrorToCommonError(response)) || (!transformErrorToCommonError&& isSdkError(response))) {
    throw new Error(`${response.errorCode}, message: ${response.message ?? 'None'}`);
```

I know that in practice this may seem a bit silly because all the property names are the same but treating two different types as the same seems like it will eventually lead to trouble.

Reply:
Totally valid, and thinking about how to improve this led to a design I like a lot better. I changed this line to instead read:

`isResponseAReportableError?: (response: unknown) => response is { errorCode: number | string; message?: string },`

Essentially, "Is the thing you got back something that has an errorCode and a message on it in a format that you accept as valid?"

One of the really nice things about doing it this way is that people can pass their functions like `isInvokeError` in to this directly now instead of wrapping it in a awkward wrapper function that lies about it being an SdkError. Handshakes all around!